reinstall-cli/src/users.rs

```diff
@@ -0,0 +1,106 @@
+use anyhow::{Context, Result};
+use std::fmt::Display;
+use std::fmt::Formatter;
+use uzers::os::unix::UserExt;
+
+#[derive(Clone, Debug)]
+pub(crate) struct UserKeys {
+    pub(crate) user: String,
+    pub(crate) authorized_keys: String,
+    pub(crate) authorized_keys_path: String,
+}
+
+impl UserKeys {
+    pub(crate) fn num_keys(&self) -> usize {
+        self.authorized_keys.lines().count()
+    }
+}
+
+impl Display for UserKeys {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "User {} ({} authorized keys)",
+            self.user,
+            self.num_keys()
+        )
+    }
+}
+
+/// # Safety
+/// See safety information in [`uzers::all_users`]. We are not impacted because we're not
+/// multi-threaded.
+pub(crate) fn get_all_users_keys() -> Result<Vec<UserKeys>> {
+    #[allow(unsafe_code)]
+    let iter = unsafe { uzers::all_users() };
+
+    let mut result = Vec::new();
+
+    for user_info in iter {
+        if user_info.uid() < 1000 && user_info.uid() != 0 {
+            tracing::debug!(
+                "Skipping user {} with uid {} because it's a system user (uid < 1000)",
+                user_info.name().to_string_lossy(),
+                user_info.uid()
+            );
+            continue;
+        }
+
+        if ["/bin/false", "/bin/true", "nologin"]
+            .into_iter()
+            .any(|x| user_info.shell().to_string_lossy().contains(x))
+        {
+            tracing::debug!(
+                "Skipping user {} with shell {} because it can't login",
+                user_info.name().to_string_lossy(),
+                user_info.shell().to_string_lossy()
+            );
+            continue;
+        }
+
+        let home_dir = user_info.home_dir();
```

Comment:
Not that it *really* matters here, but just commenting here because in the general case reading/writing a user's home directory from a privileged process requires some care, as it's otherwise easy for a user to trick root into following a symlink and reading/writing outside of it. (Really more generally having a higher privileged process read/write to a directory which is writable by a lower privileged process)

The strongest and most portable way is to fork a helper process that's `setuid` to the user, and in this case have it do the operations.

In our case since we're just reading (not writing here) we can address the symlink problem using cap-std, which we use pervasively in this project anyways.